server/src/main/java/org/candlepin/policy/js/entitlement/AbstractEntitlementRules.java

```diff
@@ -321,56 +326,82 @@ private void postUnbindVirtLimit(PoolHelper postHelper,
+            List<Pool> subPoolsForStackIds) {
+        Set<String> stackIdsThathaveSubPools = new HashSet<String>();
+        if (CollectionUtils.isNotEmpty(subPoolsForStackIds)) {
+            for (Pool pool : subPoolsForStackIds) {
+                stackIdsThathaveSubPools.add(pool.getSourceStackId());
+            }
+        }
 
         log.debug("Running virt_limit post-bind.");
-        boolean hostLimited = attributes.containsKey("host_limited") &&
-            attributes.get("host_limited").equals("true");
-        if (!c.getType().isManifest() &&
-            !"true".equalsIgnoreCase(c.getFact("virt.is_guest")) &&
-            (config.getBoolean(ConfigProperties.STANDALONE) || hostLimited)) {
-            Product product = pool.getProduct();
-            String virtLimit = attributes.get("virt_limit");
 
-            String stackId = attributes.get("stacking_id");
-            boolean createSubPool = stackId == null ||
-                poolCurator.getSubPoolForStackId(c, stackId) == null;
-
-            if (createSubPool) {
-                log.debug("Creating a new sub-pool.");
-                try {
-                    int virtQuantity = Integer.parseInt(virtLimit);
-                    if (virtQuantity <= 0) {
-                        return;
+        boolean consumerFactExpression = !c.getType().isManifest() &&
+                !"true".equalsIgnoreCase(c.getFact("virt.is_guest"));
+
+        boolean isStandalone = config.getBoolean(ConfigProperties.STANDALONE);
+
+        List<Pool> createHostRestrictedPoolFor = new ArrayList<Pool>();
+        List<Entitlement> decrementHostedBonusPoolQuantityFor = new ArrayList<Entitlement>();
+
+        for (Entitlement entitlement : entitlementMap.values()) {
+            Pool pool = entitlement.getPool();
+            Map<String, String> attributes = attributeMaps.get(pool.getId());
+            boolean hostLimited = attributes.containsKey("host_limited") &&
+                    attributes.get("host_limited").equals("true");
+
+            if (consumerFactExpression && (isStandalone || hostLimited)) {
+                String virtLimit = attributes.get("virt_limit");
+                String stackId = attributes.get("stacking_id");
+                boolean createSubPool = stackId == null || !stackIdsThathaveSubPools.contains(stackId);
+                if (createSubPool) {
+                    log.debug("Creating a new sub-pool for {}", pool);
+                    try {
+                        int virtQuantity = Integer.parseInt(virtLimit);
+                        if (virtQuantity <= 0) {
+                            continue;
+                        }
                     }
-                }
-                catch (NumberFormatException nfe) {
-                    if (!"unlimited".equals(virtLimit)) {
-                        return;
+                    catch (NumberFormatException nfe) {
+                        if (!"unlimited".equals(virtLimit)) {
+                            continue;
```

Comment:
Does `continue` work inside a `catch`?  I couldn't find any documentation indicating that it would break out of the catch block.